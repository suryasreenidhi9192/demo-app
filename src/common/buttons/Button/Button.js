import React from 'react'
import PropTypes from 'prop-types';
import { curryWithPressed } from '../../../common/shared/utils/events'

import './Button.scss'

/**
 * @public
 * A component to create HTML button elements that adds touch mutations for interaction indicators via css
 * @param  {Node}       props.children  React Component or string to render in the button/link
 * @param  {String}     props.type      Sets the html button's attribute type `button||reset||submit`
 * @param  {Boolean}    props.disabled  Disables the button via html attribute
 * @param  {Function}   props.onClick   Fires curryWithPressed util wrapping this callback,
                                        setting a 'pressed' class on the target while pressed
 * @param  {String}     props.size      Sets passed string as a concatenated classname defaults to `normal`
 * @param  {String}     props.display   Sets passed string as a concatenated classname defaults to `primary`
 * @return {Node}                       React Component
*/

const Button = ({
    children,
    type,
    disabled,
    onClick,
    size = 'normal',
    display = 'primary'
}) => {

    const extraProps = {}
    if (!!disabled) extraProps.disabled = 'disabled'
    const disabledClass = (!!disabled) ? 'disabled' : ''
    const cls = `form-button button ${size} ${display} ${disabledClass}`

    return (

        <button
            type={ type }
            className={ cls }
            onClick={ curryWithPressed(onClick) }
            {...extraProps}
        >
            { children }
        </button>

    )

}

Button.propTypes = {
    /** React Component or string to render in the button/link */
    children: PropTypes.node,
    /** Disables the button via html attribute */
    disabled: PropTypes.bool,
    /** Sets the html button's attribute type `button||reset||submit` */
    type: PropTypes.string,
    /** Fires curryWithPressed util wrapping this callback,
    setting a 'pressed' class on the target while pressed */
    onClick: PropTypes.func,
    /** Sets passed string as a concatenated classname */
    size: PropTypes.string,
    /** Sets passed string as a concatenated classname */
    display: PropTypes.string
}

export default Button
