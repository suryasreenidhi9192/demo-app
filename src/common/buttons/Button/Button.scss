@import "../../shared/css/variables";

.form-button {
    width: 100%;
    min-width: 180px;
    height: 48px;
    padding: 0 24px;
    font-family: $reg;
    font-size: 18px;
    color: var(--text);
    text-align: center;
    border: 1px solid var(--action);
    border-radius: 4px;
    box-shadow: var(--box-shadow-none);
    box-sizing: border-box;
    cursor: pointer;
    user-select: none;

    &.small {
        margin-top: 12px;
        width: 100%;
        min-width: 120px;
        height: 36px;
        padding: 0 16px;
        font-size: 15px;
    }

    &.primary {
        background-image: linear-gradient(rgba(black, 0), rgba(black, 0.2));
        color: var(--button-text);
        background-color: var(--action);
        transition: background-color 0.3s, box-shadow 0.3s, color 0.3s, border 0.3s;

        &:focus, &:hover {
            background-color: var(--action-hover);
            border: solid 1px var(--action-hover);
            box-shadow: var(--box-shadow-far);
        }

        &:active {
            background-color: var(--action-active);
            box-shadow: var(--box-shadow-none);
            border: solid 1px var(--action-active);
            transition: background-color 0.05s, box-shadow 0.05s, color 0.05s, border 0.05s;
        }

    }

    &.secondary {
        color: var(--action);
        background: var(--content-background);
        background-image: linear-gradient(rgba(black, 0), rgba(black, .1));
        transition: box-shadow 0.3s, color 0.3s, border 0.3s;

        &:focus, &:hover {
            color: var(--action-hover);
            border: solid 1px var(--action-hover);
            box-shadow: var(--box-shadow-far);
        }

        &:active {
            color: var(--action-active);
            box-shadow: var(--box-shadow-none);
            border: solid 1px var(--action-active);
            transition: box-shadow 0.05s, color 0.05s, border 0.05s;
        }
    }

    &.disabled {
        box-shadow: none;
        pointer-events: none;
        opacity: .5;
    }

}


