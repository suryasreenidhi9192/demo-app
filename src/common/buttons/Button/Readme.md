### Simple Primary Button
```jsx
    <Button> Click Here</Button>
```

### Disabled Button
```jsx
    <Button disabled> Click Here</Button>
```

### Simple Secondary Button
```jsx
    <Button display='secondary'> Click Here</Button>
```
