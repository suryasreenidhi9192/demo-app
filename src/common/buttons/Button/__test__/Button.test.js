import React from 'react'
// import { mount } from 'enzyme';
import { matchSnapshot, testComponentOnClick, snapshotIsMissingSelector } from '../../../../../__test__/__utils__'

import Button from '../'

const props = {
    onClick: jest.fn(),
    children: (<span>Click Here, Ron ;)</span>),
    type: 'submit',
    disabled: false
}

const disabledProps = {
    ...props,
    disabled: true
}

const customProps = {
    ...props,
    size: 'arbitrary',
    display: 'arbitrary'
}


describe('Button:.......................SnapShot Testing:', () => {
    test(
        'Button Matches jest snapshot',
        matchSnapshot(Button, props)
    )
    test(
        'Disabled Button Matches jest snapshot',
        matchSnapshot(Button, disabledProps)
    )
    test(
        'Button with custom size props doesn\'t match default',
        snapshotIsMissingSelector(Button, customProps, '.normal')
    )
    test(
        'Button with custom display props doesn\'t match default',
        snapshotIsMissingSelector(Button, customProps, '.primary')
    )
})

describe('Button:.......................Event Handling:', () => {
    test(
        'Clicking Button fires the handler',
        testComponentOnClick(Button, props)
    )
})
