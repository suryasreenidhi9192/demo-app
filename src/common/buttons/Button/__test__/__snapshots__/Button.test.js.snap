// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button:.......................SnapShot Testing: Button Matches jest snapshot 1`] = `
<button
  className="form-button button normal primary "
  onClick={[Function]}
  type="submit"
>
  <span>
    Click Here, Ron ;)
  </span>
</button>
`;

exports[`Button:.......................SnapShot Testing: Disabled Button Matches jest snapshot 1`] = `
<button
  className="form-button button normal primary disabled"
  disabled="disabled"
  onClick={[Function]}
  type="submit"
>
  <span>
    Click Here, Ron ;)
  </span>
</button>
`;
